### ISSUE TITLE
hdhomerun forum bug and feature request reply left.

### ISSUE THREAD

**Issue by wazerstar** [djp952/hdhomeruntray#60]
https://forum.silicondust.com/forum/viewtopic.php?p=381905#p381905

**Comment by djp952** [comment]
In case the forum link goes away, the enhancement request is to add a couple device manipulation functions:

- Ability to unlock a tuner (suggestion is to allow for clicking on the tuner status dot)
- Ability to reboot a tuner (suggestion is a clickable link in the tuner device detail pop-up footer)

An unhandled exception bug was also noted in the forum post, this has been addressed for the final release via https://github.com/djp952/hdhomeruntray/commit/51bd553a914b0bf61fc3e752efcfa11d2f54b7de

**Comment by djp952** [comment]
Per conversation on SD forums, I think this is a bad idea for general use as both options manipulate the tuner state which is outside the scope of the application.  There are also negative implications for what I feel are the lionshare of users out there that use front-ends like HDHomeRun VIEW, Kodi, Plex, etc., in that unlocking a tuner makes it eligible for re-use but does not notify the front-end of this change.  Bad things can happen, even moreso with DVR solutions that are trying to record a stream and got 'bumped' without notice.

That said, if the underlying tuner devices concerns you need to address can't be resolved, I would be willing to add a 'secret' option into the application configuration file.  It would be documented, but not accessible from the application itself.  The config file is a little buried, for example:

C:\Users\{you}\AppData\Local\Zukisoft\hdhomeruntray.exe_StrongName_bmdhf3kef35qebhzngjzmvswxmp1ztx5\0.9.9.0\**user.config**

but it's just XML and the setting would be present, just set to "false".  It would be treated like any other setting in that upgrades would keep what you had, etc, etc.  So, less of a "secret" setting than a "hidden" one I guess?

**Comment by wazerstar** [comment]
Replied on SD forums, I love it, easter egg coming :)?

**Comment by djp952** [comment]
You know me too well.  Making it secret also means it doesn’t necessarily need to look fancy, I’m in.  I already started. I just need to not screw up my own systems in the process, testing will need to occur when nothing of value is going on 😊 From: wazerSent: Friday, March 11, 2022 9:22 PMTo: djp952/hdhomeruntrayCc: Michael Brehm; CommentSubject: Re: [djp952/hdhomeruntray] hdhomerun forum bug and feature request reply left. (Issue #60) Replied on SD forums, I love it, easter egg coming :)?—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>

**Comment by wazerstar** [comment]
Bring it on, I'm not shy to test

👍 💯 🥇

**Comment by djp952** [comment]
Still working this, the UI has been the main sticking point.  I mean, it has to at least look reasonable, right?  What I've come up with is to (if you find the secret option and turn it on) is to add a toggle button in the tuner device footer.  If you toggle that, the idea is to add a panel of buttons above the footer with the "tools":

![Untitled](https://user-images.githubusercontent.com/706055/158510664-8342eaeb-4b2c-4cf9-8aab-61cc7033db70.png)

Clearly it's still a mess. The idea was that the things that say "hi" here would be more like "Unlock Tuner X", and I've found some crazy weird compositing things that need to be dealt with.  I am doing my best, it's just taking a LOT longer than I thought it would.

**Comment by djp952** [comment]
Getting closer, visually.  This has exposed some type of problem that needs to be fixed if the tuner device form resizes after it's already visible, but I'll get there.  Are you OK with just a series of simple buttons that will appear above the footer to do what you want?  The mouse was over "Unlock Tuner 0" in this screenshot, so it's highlighted:

![Untitled](https://user-images.githubusercontent.com/706055/158729243-a5881908-3806-4f51-9560-b2d015a09fac.png)

The buttons aren't actually hooked up yet, the UI for this application has been the most significant set of challenges based on the original requirements so I had to work that out first (side note, the old Windows Forms Designer added a phantom column on me and that's why the button sizes were so jacked up yesterday).  I don't foresee a problem with making the buttons do what they say they will do.

**Comment by wazerstar** [comment]
Hey, can I ask why it would not just be easier to either click the left tuner icon OR the tuner X to do the cmd instead of adding extra buttons etc :)?

Basically like the picture I made on the forums, wouldn't that be easier and more clean look for you to work with?

**Comment by djp952** [comment]
I considered the ramifications of somebody turning this on (which you know they will) and having an errant click of the mouse screw up their DVR or Kodi or <insert app name here>.  People tend to get angry about things like that 😊 “In your face and obvious” is always better! From: wazerSent: Thursday, March 17, 2022 8:54 AMTo: djp952/hdhomeruntrayCc: Michael Brehm; CommentSubject: Re: [djp952/hdhomeruntray] hdhomerun forum bug and feature request reply left. (Issue #60) Hey, can I ask why it would not just be easier to either click the left tuner icon OR the tuner X to do the cmd instead of adding extra buttons etc :)?Basically like the picture I made on the forums, wouldn't that be easier and more clean look for you to work with?—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>

**Comment by djp952** [comment]
Reboot works.  As to the "unlock" request, do you really want/need it to also de-tune the tuner?  issuing a "set /tunerX/channel none" would REALLY mess up things for most people; can you live with it just sending the "set /tunerX/lockkey force"?

**Comment by wazerstar** [comment]
> Reboot works. As to the "unlock" request, do you really want/need it to also de-tune the tuner? issuing a "set /tunerX/channel none" would REALLY mess up things for most people; can you live with it just sending the "set /tunerX/lockkey force"?

Nice, progress going :)

lockkey force will not free up the tuner, since that is what occurs sometime it will be locked until I actually reboots the entire devices, so /tunerX/channel none is really appreciated to get in and done right away after lockkey

**Comment by djp952** [comment]
In this case I feel that "Unlock Tuner X" is not a proper label to display?  The requested operation is more analogous to a "Kill Tuner X" operation.  Suggestions as to how to phrase such an operation without the word "Kill"?

**Comment by wazerstar** [comment]
> In this case I feel that "Unlock Tuner X" is not a proper label to display? The requested operation is more analogous to a "Kill Tuner X" operation. Suggestions as to how to phrase such an operation without the word "Kill"?

Perhaps Free/Release tuner

**Comment by djp952** [comment]
OK, @wazerstar, this is done and ready for you to play around with, test builds here:

https://1drv.ms/u/s!AgEGEEVzGNq-jLZqGcNCK8X1tEVG2w?e=YiIp3S

**SPECIAL INSTALLATION INSTRUCTIONS** - this is still version 0.9.9 so there are going to be a couple manual steps:

1. Uninstall the application manually from Programs and Features (or Add/Remove Programs)
2. In Explorer, navigate to **%LOCALAPPDATA%\Zukisoft\hdhomeruntray.exe_StrongName_bmdhf3kef35qebhzngjzmvswxmp1ztx5**
3. Delete the **0.9.9.0** folder
4. Install new build

The settings file has to be removed since there is a new setting but the primary version # didn't change.

**ENABLING TUNER TOOLS**

1. Launch and close the application
2. In Explorer, navigate to **%LOCALAPPDATA%\Zukisoft\hdhomeruntray.exe_StrongName_bmdhf3kef35qebhzngjzmvswxmp1ztx5**
3. Open the **0.9.9.0** folder
4. Edit the **user.config** file
5. Change the following to **Enabled**
```
            <setting name="ShowTunerTools" serializeAs="String">
                <value>Disabled</value>
            </setting>
```
6. Open the application

Now each tuner device will have a wrench toggle in the lower right corner, that when clicked will add the tuner tools buttons:

![Screenshot 2022-03-18 154620](https://user-images.githubusercontent.com/706055/159073639-0aa422a9-43b6-4216-9b0d-32e7eb0b8b0d.png)

It's a little clunky looking, but should get the job done for you.

I'll be going out of town for a few days on Sunday, so kick the tires a bit and make sure this is what you want/need for the 1.0.0 release.  You won't need to perform any manual steps like when the version actually changes to "1.0.0", and the "Enabled" setting you hand edited will be carried along.

**Comment by djp952** [comment]
Marking as closed for version 1.0.0.  That doesn't mean I expect it to be perfect, more of a way to link the Release Notes for this changeset :)
